### PR TITLE
Save initialPackages in DB instead of memory

### PR DESCRIPTION
**Pull request by vbanthia**
Currently initialPackages are stored in Memory. This PR will save initialPackages in DB instead.

As such there won't be big difference, this PR will just increase reliability of cleanup process.

### CODE REVIEW

**Comment by sorccu** [comment]
Looks nice, but two things:

1. The package list needs to be tied to the Android version somehow, so that we don't remove something that's supposed to be there in the new version, yet isn't a system package. I'm not completely sure if this is a valid concern but I know for sure we'll never even find out if we just remove them in the first place.
2. Maybe the package cleanup should be performed when the device is inserted into the system as well, just in case it died in the middle of the previous session. Currently the cleanup would only take effect after a new successfully terminated session. Perhaps it could be done right after the worker has received the list of initial packages?

**Comment by vbanthia** [comment]
1. There is a way by which we can only store 3rd party apps using `-3` option such as, `adb shell pm list packages -3`. And even while removing apps, we will take difference from current 3rd party apps, so that we won't touch system apps. But current `adbkit` does not have api to get only 3rd party apps. If this looks good, I will add that api in `adbkit`

2. I will think about how to do this.

**Comment by sorccu** [comment]
Yes, but it's not accurate. I went through a whole bunch of phones when the feature was initially implemented. It depends on the phone, but IIRC especially on many Xperia and Samsung phones there are built-in or bundled apps on that list. They can include anything from bundled crapware to scary sounding, potentially device breaking camera-related packages. Since the software is updated by the manufacturer or carrier, anyone could add any kind of bundled apps at any point. You could argue that they're not necessary, but by removing them perhaps you free up so many resources on the device that actual user devices would run into resource limitations far quicker, or perhaps someone wants to try a specific bundled integration.

If the list was accurate, we could just remove all 3rd party packages every time and do away with the package list entirely.

**Comment by vbanthia** [comment]
yeah, In that case 3rd party app will not work.

But, your first suggestion, tying initial packages with android version also won't be 100% accurate since carrier updates or manufacturer updates don't necessarily update android version. And I don't think there is any perfect solution to find out what apps are necessary for the device which we can whitelist dynamically.

One possible solution can be by adding package whitelist feature in UI, where user (or admin) can whitelist new apps. With this feature, user (or admin) would be able to see current whitelist packages and add or remove new packages.

**Comment by gunta** [comment]
Definitely having an UI for it would be the only way to let the user know for sure if we are messing with carrier apps or otherwise.

Another feature we could combine in that UI is to actually save all the APKs the user installed on the device and let them install the apps every-time they use the device.

I'd say lets merge this and discuss the rest of all the possible concerns over a new issue.

**Comment by vbanthia** [comment]
yeah, adding UI feature now will be too much. We can do this later. For now, I will fix Simo's 2nd comment and then this PR will be good to merge.

**Comment by vbanthia** [comment]
@sorccu 

I did thought about your 2nd point (cleaning device on device insertion). Looks like it is not possible to do before whitelist feature. Reason being,

If we also clean during device insertion (or DeviceReadyMessage) then there will be no chance for user to change whitelisted apps. Because we will only save it during first DeviceIntroductionMethod, i.e, when we connect device to STF for first time.

With current situation, they can at least update important apps and then reconnect device manually.

I think, current PR is not changing STF current behaviour, and to make it better we will need some device whitelist feature on UI side which can be done in other task. What do you suggest?

**Comment by gunta** [comment]
Well it changes behaviour in a way. Until now if there was a problem you would restart the server and problem gone. Now it will not be enough with that, you will need to handle the DB manually which is a bit of pain.

We should have either an UI or a command for clearing this data at least.
It could be something like `stf db clear-initial-packages`.

**Comment by vbanthia** [comment]
@gunta 

No, it will not change the current behaviour, because whenever you restart a `provider` or `device`, `wire.InitialPackagesSaveMessage` message will be sent and it will save current packages again.

**Comment by vbanthia** [comment]
So, basically the flow will be like this,

1. Whenever a device unit starts (or restarts), device will send processor `wire.InitialPackagesSaveMessage`  with current packages and processor will save them in DB.
2. Whenever a session is over, device unit ask initialPackages to processor by `wire. InitialPackagesGetMessage` and processor will get packages from DB and send them back to device through `InitialPackagesReceivedMessage` message
3. Whenever `InitialPackagesReceivedMessage` is received, device unit will calculate package difference  and remove extra one.

So basically, this PR is only storing packages in DB which was previously stored in memory.

**Comment by sorccu** [comment]
Just to be clear, is this PR meant to fix the unreliable cleanups (like mentioned in the description), or to enable an easier implementation later? Because like you say in your comments, and after reading the whole changeset, it doesn't look like it would actually functionally change anything from the way it is now (which is to say it will still have issues cleaning up after crashed sessions).

**Comment by vbanthia** [comment]
My initial goal for this PR was just to save list of initial packages into DB instead of memory for better reliability. After your comments I tried implementing better cleanup process on session crash and found out it is difficult to design because we don't know what apps to whitelist and when?

My proposal is, we can merge this PR and next action will be,
1. Add a feature where user (or admin) can whitelist defaults apps from UI.
2. Also clean when device worker starts (for session crash)
